File: lib/journal/Memo.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// external support
#include "externals.h"
// forward declarations
#include "forward.h"
// type aliases
#include "api.h"

// the global settings
#include "Chronicler.h"
// message contents
#include "Entry.h"
// the superclass
#include "Renderer.h"
// the declaration
#include "Memo.h"


// metamethods
pyre::journal::Memo::~Memo() {}


// implementation details
void
pyre::journal::Memo::header(
    palette_type & palette, linebuf_type & buffer, const entry_type & entry) const
{
    // get the page
    auto & page = entry.page();
    // if there are no contents
    if (page.empty()) {
        // we are done
        return;
    }

    // get the notes
    auto & notes = entry.notes();
    // get the channel severity
    auto severity = notes.at("severity");
    // and its name
    auto channel = notes.at("channel");

    // attempt to get location information
    // N.B.: we only print line number and function name if we know the filename
    auto loc = notes.find("filename");
    // if it's there
    if (loc != notes.end()) {
        // extract the filename
        auto filename = loc->second;
        // make some room and turn on location formatting
        buffer << palette[severity];
        // set a maximum length for the rendered filename
        const line_type::size_type maxlen = 60;
        // get the filename size
        auto len = filename.size();
        // so that names that are longer than this maximum
        if (len > maxlen) {
            // get shortened
            buffer << filename.substr(0, maxlen / 4 - 3) << "..."
                   << filename.substr(len - 3 * maxlen / 4);
        } else {
            // otherwise, render the whole name
            buffer << filename;
        }
        // reset the buffer and add a spacer
        buffer << palette["reset"] << ":";

        // attempt to get the line number
        auto & line = notes.at("line");
        // if we know it
        if (!line.empty()) {
            // render it
            buffer << palette[severity] << line << palette["reset"] << ":";
        }

        // same with the function name
        auto & function = notes.at("function");
        // if we know it
        if (!function.empty()) {
            // render it
            buffer << palette[severity] << " " << function << palette["reset"] << ":";
        }
        // wrap up the location info
        buffer << std::endl;
    }
    // if we don't have location information
    else {
        // we show some basic channel information instead
        buffer
            // print the severity
            << palette[severity] << severity << palette["reset"] << ":" << std::endl;

        // at the correct {decor} level
        if (chronicler_t::decor() > 1) {
            // print
            buffer
                // a marker
                << palette[severity] << _headerMarker
                << palette["reset"]
                // and a warning that there is no location information
                << "location information is not available"
                // done
                << std::endl;
        }
    }

    // all done
    return;
}


void
pyre::journal::Memo::body(
    palette_type & palette, linebuf_type & buffer, const entry_type & entry) const
{
    // get the page
    auto & page = entry.page();
    // if the page is empty
    if (page.empty()) {
        // nothing to do
        return;
    }

    // get the notes
    auto & notes = entry.notes();
    // get the channel severity
    auto severity = notes.at("severity");

    // go through the lines in the page
    for (auto & line : page) {
        // and render them
        buffer << palette[severity] << _bodyMarker << palette["reset"] << palette["body"] << line
               << palette["reset"] << std::endl;
    }

    // all done
    return;
}


void
pyre::journal::Memo::footer(
    palette_type & palette, linebuf_type & buffer, const entry_type & entry) const
{
    // get the page
    auto & page = entry.page();
    // if the page is empty
    if (page.empty()) {
        // nothing to do
        return;
    }

    // get the notes
    auto & notes = entry.notes();
    // get the channel severity
    auto severity = notes.at("severity");
    // and its name
    auto channel = notes.at("channel");

    if (chronicler_t::decor() > 2) {
        // look for the application name
        auto loc = notes.find("application");
        // if it's there
        if (loc != notes.end()) {
            // get the value
            auto app = loc->second;
            // render the app name
            buffer
                // first a marker
                << palette[severity] << _footerMarker
                << palette["reset"]
                // inject the application name
                << "from application " << palette[severity] << app
                << palette["reset"]
                // done
                << std::endl;
        }

        // render the channel name and severity
        buffer
            // start things off with the marker
            << palette[severity] << _footerMarker
            << palette["reset"]
            // the channel severity and name
            << "because the " << palette[severity] << severity << palette["reset"] << " channel '"
            << palette[severity] << channel << palette["reset"]
            << "' is active"
            // done with the intro line
            << std::endl;
    }

    // now for the extra notes, if any
    if (chronicler_t::decor() > 1) {
        // build a set of the keys we have processed already
        std::set<entry_type::key_type> known { "severity", "channel",  "filename",
                                               "line",     "function", "application" };
        // go through the notes
        for (auto & [key, value] : notes) {
            // if the key is in the known pile
            if (known.find(key) != known.end()) {
                // ignore it
                continue;
            }
            // otherwise, render it
            buffer << palette[severity] << _footerMarker << palette["reset"] << palette[severity]
                   << key << palette["reset"] << ": " << palette[severity] << value
                   << palette["reset"] << std::endl;
        }
    }

    // all done
    return;
}


// end of file

File: packages/journal/Memo.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


# superclass
from .Renderer import Renderer


# the renderer or developer-facing messages
class Memo(Renderer):
    """
    The renderer of user-facing messages
    """


    # implementation details
    def header(self, palette, entry):
        """
        Generate the message header
        """
        # get the page
        page = entry.page
        # if there is nothing to render
        if not page:
            # bail
            return

        # get the notes
        notes = entry.notes
        # grab the severity
        severity = notes["severity"]

        # setup the marker
        marker = self.headerMarker

        # get the name of the file
        filename = notes["filename"]
        # consider it an indication that we have location information
        if filename:
            # initialize a buffer
            buffer = [ palette[severity] ]
            # pull the longest filename we are going to print
            maxlen = self.maxlen
            # if the filename is too long
            if len(filename) > maxlen:
                # place a leading part in the buffer
                buffer.append(filename[:maxlen//4-3])
                # add an ellipsis
                buffer.append("...")
                # and the trailing part of the filename
                buffer.append(filename[-3*maxlen//4:])
            # otherwise
            else:
                # add the filename to pile
                buffer.append(filename)
            # in any case, reset the color
            buffer.append(palette["reset"])
            # and add a spacer
            buffer.append(":")

            # get the line number
            line = str(notes["line"])
            # if it's available
            if line:
                # turn color back on
                buffer.append(palette[severity])
                # add the number to the pile
                buffer.append(line)
                # reset the color
                buffer.append(palette["reset"])
                # and add a spacer
                buffer.append(":")

            # repeat with the function name
            function = notes["function"]
            # if it's available
            if function:
                # turn color back on
                buffer.append(palette[severity])
                # add the number to the pile
                buffer.append(function)
                # reset the color
                buffer.append(palette["reset"])
                # and add a spacer
                buffer.append(":")

            # assemble the line and make it available
            yield ''.join(buffer)
        # if we don't have location information
        else:
            # just print the channel severity
            buffer = [
                palette[severity],
                severity,
                palette["reset"],
                ":",
            ]
            # assemble the line and make it available
            yield ''.join(buffer)

        # all done
        return


    def body(self, palette, entry):
        """
        Generate the message body
        """
        # get the page
        page = entry.page
        # if there's nothing to do
        if not page:
            # move on
            return

        # get the notes
        notes = entry.notes

        # get the marker
        marker = self.bodyMarker
        # and the severity
        severity = notes["severity"]

        # go through the message content
        for line in page:
            # render
            buffer = [
                palette[severity], marker, palette["reset"],
                palette["body"], line, palette["reset"]
                ]
            # assemble and push
            yield ''.join(buffer)

        # all done
        return


    def footer(self, palette, entry):
        """
        Generate the message body
        """
        # get the page
        page = entry.page
        # if there's nothing to do
        if not page:
            # move on
            return

        # get the notes
        notes = entry.notes
        # grab the severity
        severity = notes["severity"]
        # and the name of the channel
        channel = notes["channel"]

        # set up the marker
        marker = self.footerMarker
        # grab the chronicler
        from . import chronicler

        # if the {chronicler} decoration level is sufficiently high
        if chronicler.decor > 2:
            # look up the application name
            app = notes.get("application")
            # if we know it
            if app:
                # render it
                buffer = [
                    palette[severity], marker, palette["reset"],
                    "from application ",
                    palette[severity], app, palette["reset"],
                ]
                # assemble and push
                yield ''.join(buffer)

            # render the channel info
            buffer = [
                palette[severity], marker, palette["reset"],
                "because the ",
                palette[severity], severity, palette["reset"],
                " channel '",
                palette[severity], channel, palette["reset"],
                "' is active"
                ]
            # assemble and push
            yield ''.join(buffer)

        # now, for the rest of the metadata
        if chronicler.decor > 1:
        # make a pile of the information we have displayed # already
            done = {
                "severity", "channel", "application",
                "filename", "line", "function", "source",
            }
            # go through the metadata
            for key, value in notes.items():
                # if this is something we've dealt with
                if key in done:
                    # skip it
                    continue
                # otherwise, render it
                buffer = [
                    palette[severity], marker, palette["reset"],
                    palette[severity], key, palette["reset"],
                    ": ",
                    palette[severity], value, palette["reset"],
                    ]
                # assemble and push
                yield ''.join(buffer)

        # all done
        return


    # implementation details
    maxlen = 60
    headerMarker = " >> "
    bodyMarker = " -- "
    footerMarker = " .. "


# end of file
